src/core/drude.hpp

```diff
@@ -0,0 +1,110 @@
+
+/*
+  Copyright (C) 2010,2012,2013 The ESPResSo project
+  Copyright (C) 2002,2003,2004,2005,2006,2007,2008,2009,2010 
+  Max-Planck-Institute for Polymer Research, Theory Group
+  
+  This file is part of ESPResSo.
+  
+  ESPResSo is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+  
+  ESPResSo is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+  
+  You should have received a copy of the GNU General Public License
+  along with this program.  If not, see <http://www.gnu.org/licenses/>. 
+*/
+
+#ifndef DRUDE_H
+#define DRUDE_H
+/** \file drude_correction.h
+ *  Routines to subtract the electrostatic Energy and/or the electrostatic force 
+ *  for a particle pair and correct the equations of motion for a Drude pair.
+ *  \ref forces.c
+*/
+
+#include "utils.hpp"
+#include "interaction_data.hpp"
+#include "integrate.hpp"
+#include "random.hpp"
+#include "p3m.hpp"
+
+#ifdef DRUDE
+
+// Set the parameters for the Drude correction
+int drude_set_params(int bond_type, double temp_com, double gamma_com, double temp_drude, double gamma_drude, double k, double r_cut);
+
+/** Computes the necessary corrections for the properly simulation of Drude Particles
+    and adds this force to the particle forces. 
+    @param p1        Pointer to first particle.
+    @param p2        Pointer to second/middle particle.
+    @param iaparams  Parameters of interaction
+    @param dx        change in position
+    @param force1 and force2     force on particles
+    @return true if bond is broken
+*/
+
+inline int calc_drude_forces(Particle *p1, Particle *p2, Bonded_ia_parameters *iaparams, double dx[3], double force1[3], double force2[3])
+{
+  //Bond broke?
+  double dist = Utils::veclen(dx);
+  if (iaparams->p.drude.r_cut > 0.0 && dist > iaparams->p.drude.r_cut) {
+    return 1;
+  }
+
+  double force_harmonic, force_lv_com, force_lv_dist, com_vel, dist_vel;
+  double chgfac = p1->p.q*p2->p.q;
+  double fac_harmonic = -iaparams->p.drude.k;
+  double mass_tot_inv = 1.0 / (p1->p.mass + p2->p.mass);
+
+  double force_p3m_sr[3] = {0,0,0}; 
+
+  p3m_add_pair_force(chgfac, dx, dist*dist, dist, force_p3m_sr);
+
+  for (int i=0; i<3; i++) {
+
+      //Langevin thermostat for center of mass
+      com_vel = mass_tot_inv * (p1->p.mass * p1->m.v[i] + p2->p.mass * p2->m.v[i]);
+      force_lv_com =  -iaparams->p.drude.gamma_com / time_step * com_vel + sqrt(2.0 * iaparams->p.drude.gamma_com / time_step * iaparams->p.drude.temp_com) * gaussian_random_cut();
```

Comment:
How sure are you about this numbers? I think for standard langevin is `sqrt(24.0 * kT / time_step * gamma)`. Also I don't think you have to use gaussian noise.

Please add tests for the respective velocity distributions, like they exist for the other thermostats.

Reply:
The velocity distributions have been tested, the factor 2 and gaussian noise (variance 1) are correct, 24 is for uniform random numbers (variance 1/12) . But I will try to use uniform rnd for performance and add tests.

Reply:
Ah yes, you're right. But I think you should use uniform noise.